Use Velox Parquet Writer in ParquetTpchTest

duanmeng opened the pull request:
The current implementation uses the DuckDB writer. We must use the Parquet
writer in Velox. This PR does the following,
- Get off the DuckDB's TPC-H extension and its writer to write
data files (`COPY SELECT TO PATH`).
- Use the Velox TPC-H connector to generate data.
- Use the generated data to create DuckDB tables used by
the `DuckDbQueryRunner` (aka expectedQueryRunner).
- Use ParquetWriter to write the same generated data to the table path
used by the `TpchQueryBuilder`.

Details in https://github.com/facebookincubator/velox/issues/6495

duanmeng commented:
@majetideepak Hi Deepak, I've added a parquet writer in ParquetTpchTest as we discussed in https://github.com/facebookincubator/velox/issues/6495, could you please help to review? Thanks a lot.

majetideepak commented on velox/dwio/parquet/tests/ParquetTpchTestBase.h:
```diff
@@ -73,28 +75,65 @@ class ParquetTpchTestBase : public testing::Test {
+      // related to sorting, and should be updated with a more proper name.
+      auto result = duckDb_->executeOrdered(query, tableSchema);
+      auto ptr = velox::test::VectorMaker::rowVector(tableSchema, result, pool);
+      result.size();
```
dummy statement?

duanmeng replied:
`DuckDbQueryRunner->executeOrdered` just executes the query and materializes the result, it does not do any sort. But the API
`DuckDbQueryRunner->execute` sort the result using `multiset`. I want to update the API name of `DuckDbQueryRunner->executeOrdered`, maybe remove the *ordered*. So I added a todo comment, should I remove it?

duanmeng replied:
@majetideepak Sorry, I saw the wrong place, I will remove the `result.size()` and update the var names.

majetideepak commented on velox/dwio/parquet/tests/ParquetTpchTestBase.h:
```diff
@@ -127,53 +166,95 @@ class ParquetTpchTestBase : public testing::Test {
+                      l_partkey,
+                      l_suppkey,
+                      l_linenumber,
+                      L_QUANTITY::DOUBLE AS quantity,
```
Why cast to Double?

majetideepak replied:
Thanks for clarifying. Can we remove the cast to double? We now support decimal types.

duanmeng replied:
Got it, I've removed all the `cast to double` and rerun the tpch test, all passed.

duanmeng commented on velox/dwio/parquet/tests/ParquetTpchTestBase.h:
```diff
@@ -73,28 +75,65 @@ class ParquetTpchTestBase : public testing::Test {
+
+      auto scanPlan = PlanBuilder().values({ptr});
+      auto plan = scanPlan
+                      .tableWrite(
```
Got it, updated. It is much simpler now, thanks.

majetideepak commented on velox/dwio/parquet/tests/ParquetTpchTestBase.h:
```diff
@@ -127,53 +146,95 @@ class ParquetTpchTestBase : public testing::Test {
+    TpchQueryBuilder(dwio::common::FileFormat::PARQUET);
+
+std::unordered_map<std::string, const std::string>
+    ParquetTpchTestBase::kDuckDbParquetWriteSQL_ = {
```
@duanmeng can we remove this map now that the casts are gone? That was my initial goal. Thanks!

duanmeng replied:
Do you mean remove this map and use `auto query = fmt::format("SELECT * FROM {}", tableName);`?

duanmeng replied:
>"SELECT * FROM

@majetideepak I just tested it, and all tests passed.

duanmeng commented:
@majetideepak I've resolved all your comments according to your comments, could you please help to take a look? Thanks a lot.

majetideepak commented on velox/dwio/parquet/tests/ParquetTpchTestBase.h:
```diff
@@ -73,28 +75,44 @@ class ParquetTpchTestBase : public testing::Test {
+
+      auto tableSchema = velox::tpch::getTableSchema(table);
+      auto query = fmt::format("SELECT * FROM {}", tableName);
+      // TODO: The name of API `executeOrdered` is confused, it do nothing
```
We can remove this TODO. Probably, add an issue in Velox.

duanmeng replied:
Fixed

majetideepak commented on velox/vector/tests/utils/VectorMaker.cpp:
```diff
@@ -27,6 +44,41 @@ std::shared_ptr<const RowType> VectorMaker::rowType(
   return ROW(std::move(names), std::move(types));
 }
 
+// static
```
remove?

duanmeng replied:
Done

majetideepak commented on velox/vector/tests/utils/VectorMaker.h:
```diff
@@ -54,6 +54,11 @@ class VectorMaker {
   static std::shared_ptr<const RowType> rowType(
       std::vector<std::shared_ptr<const Type>>&& types);
 
```
add comment ` /// Create an RowVector from std::vector<std::vector<variant>>`
Should we follow `arrayOfRowVector()` implementation and make this a non-static function?

duanmeng replied:
Ok, fixed. Add `std::shared_ptr<velox::test::VectorMaker> vectorMaker_` in `ParquetTpchTestBase`, and make this function a non-static function.

majetideepak reviewed (commented):
@duanmeng some minor comments. Looks great overall.

duanmeng commented:
@majetideepak I've resolved all the comments according to your guidance, could you help to take a look? Thanks a lot.

majetideepak reviewed (approved):
Thanks, @duanmeng

majetideepak commented:
@pedroerp Can you please review and help merge? This PR is needed to remove the dependency on Duck Parquet amalgamation.

pedroerp commented on velox/dwio/parquet/tests/ParquetTpchTestBase.h:
```diff
@@ -73,28 +76,41 @@ class ParquetTpchTestBase : public testing::Test {
-      duckDb_->execute(query);
+
+      auto tableSchema = velox::tpch::getTableSchema(table);
+      auto query = fmt::format("SELECT * FROM {}", tableName);
```
not sure I follow. Why do we still need duckDB here? Couldn't we just generate the TPC-H data using TpchConnector and write them to a parquet file?

majetideepak replied:
If we write them to parquet, we need to enable the parquet extension in DuckDB to read which we want to avoid.
The tpch connector results in Velox are also not matching the tpch results in Duck since the date generated is different.
This is the simpler workaround. We use the same date from duck tpch extension, write to parquet, and compare velox tpch plans on parquet against duck tpch extension.

majetideepak replied:
@duanmeng Pedro and I discussed offline and we felt it is better to get off the Duck tpch extension.
I added an API to `tpch::getQuery()` here https://github.com/facebookincubator/velox/pull/6784
Can we go back to the Velox tpch connector approach to generate data? Then we can save the tables in Duck via `DuckDbQueryRunner::createTable`.

duanmeng replied:
OK, got it. I will update this PR once https://github.com/facebookincubator/velox/pull/6784 lands.

pedroerp replied:
Thank you guys

majetideepak replied:
@duanmeng the PR landed. Thanks!

duanmeng replied:
@majetideepak @pedroerp  I've updated this PR to use tpch connector approach.

pedroerp reviewed (commented):
Thanks for looking into this! A few inline questions/comments.

duanmeng commented on velox/dwio/parquet/tests/ParquetTpchTestBase.h:
```diff
@@ -73,28 +76,41 @@ class ParquetTpchTestBase : public testing::Test {
-    const auto tableNames = tpchBuilder_.getTableNames();
-    for (const auto& tableName : tableNames) {
+  static void saveTpchTablesAsParquet(memory::MemoryPool* pool) {
+    static constexpr auto tables = {
```
Got it, done.

duanmeng commented on velox/dwio/parquet/tests/ParquetTpchTestBase.h:
```diff
@@ -73,28 +76,41 @@ class ParquetTpchTestBase : public testing::Test {
+        tpch::Table::TBL_REGION};
+
+    for (const auto& table : tables) {
+      auto tableName = std::string(toTableName(table));
```
Fixed

duanmeng commented:
@majetideepak @pedroerp I've updated the PR to use tpch connector to generate data and save the tables in Duck via `DuckDbQueryRunner::createTable`, could you help to take a look?

majetideepak commented on velox/dwio/parquet/tests/ParquetTpchTestBase.h:
```diff
@@ -74,28 +81,36 @@ class ParquetTpchTestBase : public testing::Test {
 
-  void TearDown() {
+  static void TearDownTestSuite() {
     connector::unregisterConnector(kHiveConnectorId);
```
add `connector::unregisterConnector(kTpchConnectorId);`

majetideepak reviewed (approved):
@duanmeng, Thanks!

majetideepak commented:
@mbasmanova, I believe Pedro is taking some time off. Can someone help review and merge this? Thanks.

majetideepak commented on velox/tpch/gen/TpchGen.cpp:
```diff
@@ -88,26 +88,26 @@ int32_t toDate(std::string_view stringDate) {
 
 } // namespace
 
-std::string_view toTableName(Table table) {
```
@duanmeng We can keep this as `std::string_view` as before.

mbasmanova commented on velox/dwio/parquet/tests/ParquetTpchTestBase.h:
```diff
@@ -29,40 +31,45 @@
   // Setup a DuckDB instance for the entire suite and load TPC-H data with scale
   // factor 0.01.
-  void SetUp() {
+  static void SetUpTestSuite() {
```
Would you move implementation to .cpp file? Same in other places.

duanmeng replied:
Got it, done.

mbasmanova commented on velox/dwio/parquet/tests/ParquetTpchTestBase.h:
```diff
@@ -74,28 +81,36 @@ class ParquetTpchTestBase : public testing::Test {
+              kTpchConnectorId, 1, 0));
+
+      auto rows = AssertQueryBuilder(plan).splits({split}).copyResults(pool);
+      duckDb_->createTable(tableName, {rows});
```
Why do we need this? Can we not create single plan like this:

```
  PlanBuilder()
    .tableScan(table, std::move(names), 0.01)
    .tableWrite(tableDirectory, dwio::common::FileFormat::PARQUET)
    .planNode()
```

duanmeng replied:
We need to use the Velox tpch connector to generate data to create DuckDB tables used by
the `DuckDbQueryRunner`.

The same generated data is also used by the `TpchQueryBuilder`.

mbasmanova commented:
@duanmeng Would you double check that PR description accurately describes the changes?

duanmeng commented:
> @duanmeng Would you double check that PR description accurately describes the changes?

@mbasmanova Hi Masha, I've updated the PR description, resolved all the comments, and moved the implementation into the `ParquetTpchTestBase.cpp`.

mbasmanova commented on velox/dwio/parquet/tests/ParquetTpchTest.cpp:
```diff
@@ -17,6 +17,8 @@
 #include "velox/common/base/tests/GTestUtils.h"
 #include "velox/dwio/parquet/tests/ParquetTpchTestBase.h"
 
+using namespace facebook::velox::exec::test;
```
Is this needed?

duanmeng replied:
Yes, `ParquetTpchTestBase` and `assertQuery` used by `ParquetTpchTest` are in this namespace.

mbasmanova commented on velox/dwio/parquet/tests/ParquetTpchTestBase.cpp:
```diff
@@ -0,0 +1,125 @@
+          ->newConnector(kHiveConnectorId, nullptr);
+  connector::registerConnector(hiveConnector);
+
+  std::shared_ptr<memory::MemoryPool> rootPool{
```
rootPool and pool variable are used only to call saveTpchTablesAsParquet; maybe move them there

duanmeng replied:
Fixed

mbasmanova commented on velox/dwio/parquet/tests/ParquetTpchTestBase.cpp:
```diff
@@ -0,0 +1,125 @@
+  }
+  tempDirectory_ = TempDirectoryPath::create();
+
+  functions::prestosql::registerAllScalarFunctions();
```
I wonder if this logic has to be in static SetupTestSuite or if it can be moved back to Setup.

majetideepak replied:
@duanmeng `SetupTestSuite` I believe cause issues when running tests in parallel. Let's keep the old `SetUp`. I don't think is an issue if we create these instances per test.

mbasmanova commented on velox/dwio/parquet/tests/ParquetTpchTestBase.cpp:
```diff
@@ -0,0 +1,125 @@
+    auto tableName = toTableName(table);
+    auto tableDirectory = fmt::format("{}/{}", tempDirectory_->path, tableName);
+    auto tableSchema = tpch::getTableSchema(table);
+    auto names = tableSchema->names();
```
nit: names -> columnNames

duanmeng replied:
Fixed

mbasmanova commented on velox/dwio/parquet/tests/ParquetTpchTestBase.cpp:
```diff
@@ -0,0 +1,125 @@
+          ->newConnector(kTpchConnectorId, nullptr);
+  connector::registerConnector(tpchConnector);
+  saveTpchTablesAsParquet(pool.get());
+  tpchBuilder_.initialize(tempDirectory_->path);
```
If we move this logic back to Setup(), we can keep tpchBuilder_ variable non-static.

mbasmanova reviewed (commented):
@duanmeng Thank you for iterating. Looks like we don't need to move all of SetUp() logic into SetUpTestSuite(), only some.

duanmeng commented on velox/dwio/parquet/tests/ParquetTpchTestBase.cpp:
```diff
@@ -0,0 +1,125 @@
+  if (duckDb_ == nullptr) {
+    duckDb_ = std::make_shared<DuckDbQueryRunner>();
+  }
+  tempDirectory_ = TempDirectoryPath::create();
```
> create duckDB_ only if it is null

This is the original code. But I think we should create it unconditionally. I will remove the `if condition`.

majetideepak replied:
We can remove the if condition.

duanmeng replied:
Done

majetideepak commented:
@duanmeng, we can move everything to `ParquetTpchTest.cpp`. We created the ParquetTpchTestBase.h earlier to support a similar ParquetTpchTest.cpp under the DuckDB Parquet reader. But the Duck Parquet reader has been removed. So we do not need the base anymore.

majetideepak commented:
> Looks like we don't need to move all of SetUp() logic into SetUpTestSuite(), only some.

@mbasmanova GTest does not allow both `void SetUp() override` and `static void SetUpTestSuite()` in the same test.
We can only have one. Let's keep the old `void SetUp() override` to enable parallel tests.

duanmeng commented:
> @duanmeng, we can move everything to `ParquetTpchTest.cpp`. We created the ParquetTpchTestBase.h earlier to support a similar ParquetTpchTest.cpp under the DuckDB Parquet reader. But the Duck Parquet reader has been removed. So we do not need the base anymore.

@majetideepak Got it, I've removed `ParquetTpchTestBase`, moved everything to `ParquetTpchTest.cpp` and back to old `Setup` method. cc @mbasmanova

mbasmanova commented on velox/dwio/parquet/tests/ParquetTpchTest.cpp:
```diff
@@ -14,12 +14,136 @@
+            ->newConnector(kHiveConnectorId, nullptr);
+    connector::registerConnector(hiveConnector);
+
+    std::shared_ptr<memory::MemoryPool> rootPool{
```
This variable and 'pool' are no longer needed, right?

mbasmanova commented on velox/dwio/parquet/tests/ParquetTpchTest.cpp:
```diff
@@ -14,12 +14,132 @@
+
+    parse::registerTypeResolver();
+    filesystems::registerLocalFileSystem();
+    parquet::registerParquetReaderFactory();
```
Don't we need to register writer factory? CC: @majetideepak 

I'm getting errors:

```
Error Code: INVALID_STATE
Reason: WriterFactory is not registered for format parquet
Retriable: False
Expression: it != writerFactories().end()
Function: getWriterFactory
File: fbcode/velox/dwio/common/WriterFactory.cpp
Line: 55
```

majetideepak replied:
@mbasmanova You are right, we should register the Parquet Writer Factory. How was this fixed?

majetideepak replied:
I see that you fixed it internally. https://github.com/facebookincubator/velox/commit/9f9fb193e97060bb74eb71df0fb53f33995d7278
Thank you! Looks neat to leave a new line.
